### PR TITLE
feat(transport): Use a shorter ECN probe threshold initially

### PR DESCRIPTION
**Pull request by mxinden**
This commit adds a test where a client connects to a server over a connection dropping all ECN marked datagrams (ECN blackhole) in both directions, asserting 43 RTT to detect ECN blackhole, disable ECN and eventually establish connection. Test assumes 20 RTT for client to detect blackhole, 20 RTT for server to detect blackhole and 3 RTT for handshake to be confirmed.

---

Test for https://github.com/mozilla/neqo/issues/1925.

How likely are ECN blackholes on today's Internet? Should this block [ECN roll-out on Firefox](https://phabricator.services.mozilla.com/D212959)? //CC @larseggert 

[Bauer et. alia. in "Measuring the State of ECN Readiness in Servers, Clients, and Routers"](https://conferences.sigcomm.org/imc/2011/docs/p171.pdf) suggests that ECN blackholes are rare.

<details>
  <summary>Test output</summary>
  
  
  ```
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Client at 0ms
0s  0ms DEBUG ECN probing: sent 1 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 50ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 100ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 150ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 200ms
0s  0ms DEBUG ECN probing: sent 2 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 250ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 300ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 350ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 400ms
0s  0ms DEBUG ECN probing: sent 3 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 450ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 500ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 550ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 600ms
0s  1ms DEBUG ECN probing: sent 4 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 650ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 700ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 750ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 800ms
0s  1ms DEBUG ECN probing: sent 5 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 850ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 900ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 950ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1000ms
0s  1ms DEBUG ECN probing: sent 6 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1050ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1100ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1150ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1200ms
0s  2ms DEBUG ECN probing: sent 7 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1250ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1300ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1350ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1400ms
0s  2ms DEBUG ECN probing: sent 8 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1450ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1500ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1550ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1600ms
0s  3ms DEBUG ECN probing: sent 9 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1650ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1700ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1750ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1800ms
0s  3ms DEBUG ECN probing: sent 10 probes
0s  3ms DEBUG ECN probing concluded with 10 probes sent
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1850ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 1900ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 1950ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2000ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Init
Processing Server at 2050ms
0s  5ms DEBUG ECN probing: sent 1 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2100ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2150ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2200ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2250ms
0s  6ms DEBUG ECN probing: sent 2 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2300ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2350ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2400ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2450ms
0s  8ms DEBUG ECN probing: sent 3 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2500ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2550ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2600ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2650ms
0s  9ms DEBUG ECN probing: sent 4 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2700ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2750ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2800ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2850ms
0s 10ms DEBUG ECN probing: sent 5 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 2900ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 2950ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3000ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3050ms
0s 12ms DEBUG ECN probing: sent 6 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3100ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3150ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3200ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3250ms
0s 13ms DEBUG ECN probing: sent 7 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3300ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3350ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3400ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3450ms
0s 14ms DEBUG ECN probing: sent 8 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3500ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3550ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3600ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3650ms
0s 15ms DEBUG ECN probing: sent 9 probes
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3700ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3750ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3800ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3850ms
0s 17ms DEBUG ECN probing: sent 10 probes
0s 17ms DEBUG ECN probing concluded with 10 probes sent
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 3900ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 3950ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 4000ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 4050ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = WaitInitial
Processing Client at 4100ms
0s 18ms TRACE RTT latest=100ms -> estimate=100ms~50ms
0s 18ms WARN ECN validation failed, no ECN counts in ACK frame
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 4150ms
0s 19ms TRACE RTT latest=500ms -> estimate=500ms~250ms
0s 19ms TRACE RTT latest=100ms -> estimate=450ms~287.5ms
0s 20ms WARN ECN validation failed, no ECN counts in ACK frame
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Client at 4200ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Handshaking
Processing Server at 4250ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Connected
Processing Client at 4300ms
0s 21ms TRACE RTT latest=100ms -> estimate=100ms~37.5ms
0s 21ms TRACE RTT latest=100ms -> estimate=100ms~28.125ms
[neqo-transport/src/connection/tests/mod.rs:190:13] c.state() = Confirmed
0s 22ms TRACE RTT latest=100ms -> estimate=406.25ms~303.125ms
test connection::tests::ecn::handshake_delay_with_ecn_blackhole ... ok

  ```
</details>

### CODE REVIEW

**Comment by mxinden** [comment]
Friendly ping @larseggert.

**Comment by larseggert** [comment]
Thanks. I'll add a fix so that we stop trying ECN earlier in this case.

File: neqo-transport/src/connection/tests/mod.rs [hunk 1]
```diff
@@ -208,17 +210,19 @@ fn handshake_with_modifier(
...
             did_ping[a.role()] = true;
         }
-        assert!(had_input || output.is_some());
-        if let Some(d) = output {
-            input = modifier(d);
-        } else {
-            input = output;
-        }
+        // TODO
+        // assert!(dbg!(had_input) || dbg!(output.is_some()));
```
Comment (larseggert) [inline neqo-transport/src/connection/tests/mod.rs:224]: Can we put this back?

Comment (mxinden) [reply neqo-transport/src/connection/tests/mod.rs:224]: This will need an update. In the case of an ECN blackhole a server would neither receive an input, nor would it produce any output.

Comment (larseggert) [reply neqo-transport/src/connection/tests/mod.rs:224]: OK, we should do that.

**Comment by larseggert** [comment]
@mxinden WDYT?

File: neqo-transport/src/ecn.rs [hunk 2]
```diff
@@ -9,18 +9,29 @@ use std::ops::{AddAssign, Deref, DerefMut, Sub};
...
-use crate::{packet::PacketNumber, recovery::SentPacket};
+use crate::{
+    packet::{PacketNumber, PacketType},
+    recovery::SentPacket,
+};
 
 /// The number of packets to use for testing a path for ECN capability.
 pub const ECN_TEST_COUNT: usize = 10;
 
+/// The number of packets to use for testing the initial a path for ECN capability when exchanging
```
Comment (mxinden) [inline neqo-transport/src/ecn.rs:20]:
```suggestion
/// The number of packets to use for testing a path for ECN capability when exchanging
```

File: neqo-transport/src/ecn.rs [hunk 3]
```diff
@@ -9,18 +9,29 @@ use std::ops::{AddAssign, Deref, DerefMut, Sub};
...
+/// Initials during the handshake. This is a lower number than `ECN_TEST_COUNT` to avoid
```
Comment (mxinden) [inline neqo-transport/src/ecn.rs:21]:
```suggestion
/// Initials during the handshake. This is a lower number than [`ECN_TEST_COUNT`] to avoid
```

In case we ever change the constant name, `cargo doc` will fail if this is not updated as well. Neat trick to keep documentation up-to-date.

File: neqo-transport/src/ecn.rs [hunk 4]
```diff
@@ -9,18 +9,29 @@ use std::ops::{AddAssign, Deref, DerefMut, Sub};
...
+/// unnecessarily delaying the handshake; we would otherwise double the PTO `ECN_TEST_COUNT` times.
```
Comment (mxinden) [inline neqo-transport/src/ecn.rs:22]:
```suggestion
/// unnecessarily delaying the handshake; we would otherwise double the PTO [`ECN_TEST_COUNT`] times.
```

File: neqo-transport/src/ecn.rs [hunk 5]
```diff
@@ -9,18 +9,29 @@ use std::ops::{AddAssign, Deref, DerefMut, Sub};
...
 /// The state information related to testing a path for ECN capability.
 /// See RFC9000, Appendix A.4.
 #[derive(Debug, PartialEq, Clone)]
 enum EcnValidationState {
     /// The path is currently being tested for ECN capability, with the number of probes sent so
     /// far on the path during the ECN validation.
-    Testing(usize),
+    Testing {
+        probes_sent: usize,
+        probes_lost: usize,
```
Comment (mxinden) [inline neqo-transport/src/ecn.rs:33]:
```suggestion
        initial_probes_lost: usize,
```

These are not the probes lost in general, but the `PacketType::Initial` only, correct?

File: neqo-transport/src/ecn.rs [hunk 6]
```diff
@@ -138,6 +152,28 @@ impl EcnInfo {
...
+            probes_sent,
+            probes_lost,
+        } = &mut self.state
+        {
+            *probes_lost += lost_packets
+                .iter()
+                .filter(|p| p.packet_type() == PacketType::Initial && p.ecn_mark().is_ecn_marked())
+                .count();
+            // If we have lost all initial probes a bunch of times, we can conclude that the path
+            // is not ECN capable.
```
Comment (mxinden) [inline neqo-transport/src/ecn.rs:166]:
```suggestion
            // If we have lost all initial probes a bunch of times, we can conclude that the path
            // is not ECN capable and likely drops all ECN marked packets.
```

The path is not only not capable, but also likely drops all packets in this case, right? If so, differentiating the two might be easier for folks to comprehend.

**Review by mxinden — COMMENTED** [review]
Implementation looks good to me. Thank you!

Intuitively 6 RTT to detect an ECN black-hole on the client seems fine under the assumption that ECN black-holes are very rare.

I can clean-up the tests later today or tomorrow unless you want to.

File: neqo-transport/src/connection/tests/mod.rs [hunk 7]
```diff
@@ -213,12 +212,7 @@ fn handshake_with_modifier(
             a.test_frame_writer = None;
             did_ping[a.role()] = true;
         }
-        assert!(had_input || output.is_some());
```
Comment (mxinden) [inline neqo-transport/src/connection/tests/mod.rs:216]: As discussed before, this is no longer true with the `handshake_delay_with_ecn_blackhole` where initially a server will not receive nor send anything due to the blackhole.

File: neqo-transport/src/connection/tests/mod.rs [hunk 8]
```diff
@@ -213,12 +212,7 @@ fn handshake_with_modifier(
...
-        if let Some(d) = output {
-            input = modifier(d);
-        } else {
-            input = output;
-        }
+        input = output.and_then(modifier);
```
Comment (mxinden) [inline neqo-transport/src/connection/tests/mod.rs:215]: This is equivalent. Happy to extract into separate pull request.

**Review by mxinden — COMMENTED** [review]
@larseggert I cleaned-up the tests. Mind taking another look?

**Comment by larseggert** [comment]
@KershawChang or @martinthomson, could you take a look? This would let us enable ECN in Gecko w/@mxinden's patch.

**Comment by larseggert** [comment]
Would be nice to get this reviewed and merged.

**Review by KershawChang — APPROVED** [review]
Looks good. Thanks!
